grpc-client-server/prober_server/main.go

```diff
@@ -9,37 +9,70 @@ import (
 	"net/http"
 	"time"
 
-	pb "github.com/CodeYourFuture/immersive-go-course/grpc-client-server/prober"
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"google.golang.org/grpc"
 )
 
 var (
-	port = flag.Int("port", 50051, "The server port")
+	port         = flag.Int("port", 50051, "The server port")
+	LatencyGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "latency_gauge",
+		Help: "The latency of the requests to the endpoint",
+	}, []string{"endpoint"})
 )
 
 // server is used to implement prober.ProberServer.
-type server struct {
+type Server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
+func (s *Server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
 	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
+	totalMsecs := 0
+	failed := 0
+	for i := 0; i < int(in.GetNumberOfRequests()); i++ {
+		start := time.Now()
+		resp, err := http.Get(in.GetEndpoint())
+		if err != nil {
+			log.Printf("could not probe: %v", err)
+			failed++
+			continue
+		}
+		defer resp.Body.Close()
```

Comment:
We generally avoid defer inside of for loops - can you do some research to find out why? How can we achieve a similar effect without running into the same issues?

Reply:
Hmm interesting reads and definitely learned a few things about `defer` today:
1. It doesn't execute defer on each loop iteration so in this case body doesn't get closed
2. It doesn't execute defere for each item in the loop event when function exits, and it only executes it for the last item in the loop

Why it's bad: in this usecase every connection *could* stay open even after we are done using them which will use up resources.

Reply:
Exactly - good research!